logstash-core/src/main/java/org/logstash/secret/store/backend/JavaKeyStore.java

```diff
@@ -0,0 +1,209 @@
+        try {
+            this.keyStorePath = keyStorePath;
+            this.keyStore = KeyStore.getInstance(KEYSTORE_TYPE);
+            this.keyStorePass = SecretStoreUtil.base64Encode(keyStorePass);
+            ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+            readLock = readWriteLock.readLock();
+            writeLock = readWriteLock.writeLock();
+            this.protectionParameter = new PasswordProtection(this.keyStorePass);
+            SecretIdentifier logstashMarker = new SecretIdentifier(LOGSTASH_MARKER);
+
+            try (final InputStream is = Files.newInputStream(keyStorePath)) {
+                try {
+                    keyStore.load(is, this.keyStorePass);
+                } catch (IOException ioe) {
+                    if (ioe.getCause() instanceof UnrecoverableKeyException) {
+                        throw new SecretStoreException.AccessException(
+                                String.format("Can not access Java keystore at %s. Please verify correct file permissions and keystore password.",
+                                        keyStorePath.toAbsolutePath()), ioe);
+                    } else {
+                        throw new SecretStoreException.NotLogstashKeyStore(String.format("Found a file at %s, but it is not a valid Logstash keystore.",
+                                keyStorePath.toAbsolutePath().toString()), ioe);
+                    }
+                }
+                byte[] marker = retrieveSecret(logstashMarker);
+                if (marker == null) {
+                    throw new SecretStoreException.NotLogstashKeyStore(String.format("Found a keystore at %s, but it is not a Logstash keystore.",
+                            keyStorePath.toAbsolutePath().toString()));
+                }
+                LOGGER.debug("Using existing keystore at {}", keyStorePath.toAbsolutePath());
+            } catch (NoSuchFileException noSuchFileException) {
+                LOGGER.info("Keystore not found at {}. Creating new keystore.", keyStorePath.toAbsolutePath());
+                //wrapped in a system property mostly for testing, but could be used for more restrictive defaults
+                String keyStorePermissions = System.getProperty("logstash.keystore.file.perms", "rw-rw----");
+                //create the keystore on disk with a default entry to identify this as a logstash keystore
+                Files.createFile(keyStorePath, PosixFilePermissions.asFileAttribute(PosixFilePermissions.fromString(keyStorePermissions)));
+                try (final OutputStream os = Files.newOutputStream(keyStorePath, StandardOpenOption.WRITE)) {
+                    keyStore = KeyStore.Builder.newInstance(KEYSTORE_TYPE, null, protectionParameter).getKeyStore();
+                    SecretKeyFactory factory = SecretKeyFactory.getInstance("PBE");
+                    byte[] base64 = SecretStoreUtil.base64Encode(LOGSTASH_MARKER.getBytes(StandardCharsets.UTF_8));
```

Comment:
It seems like we're duplicating a lot of the logic from our persistSecret; would it be possible to share that logic (e.g., once this object has been constructed "enough", use its own `JavaKeyStore#persistSecret(byte[])` instead of re-implementing it in the huge constructor?

Reply:
Agree could remove some duplication here...however, there are a couple a difference , for example not base64 encoding the marker, and (soon to be) reloading the keystore (which creates a chicken/egg issue). 

I have more changes coming, and I don't like this monster constructor...but for now I will keep it as-is and look to DRY things up once the code is unlikely to change anymore.